doc/yabai.1
'\" t
.\"     Title: yabai
.\"    Author: [see the "AUTHOR(S)" section]
.\" Generator: Asciidoctor 2.0.16
.\"      Date: 2022-09-14
.\"    Manual: Yabai Manual
.\"    Source: Yabai
.\"  Language: English
.\"
.TH "YABAI" "1" "2022-09-14" "Yabai" "Yabai Manual"
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.ss \n[.ss] 0
.nh
.ad l
.de URL
\fI\\$2\fP <\\$1>\\$3
..
.als MTO URL
.if \n[.g] \{\
.  mso www.tmac
.  am URL
.    ad l
.  .
.  am MTO
.    ad l
.  .
.  LINKSTYLE blue R < >
.\}
.SH "NAME"
.sp
yabai
.SH "SYNOPSIS"
.sp
\fByabai\fP [\fB\-v\fP,\fB\-\-version\fP|\fB\-V\fP,\fB\-\-verbose\fP|\fB\-m\fP,\fB\-\-message\fP \fImsg\fP|\fB\-c\fP,\fB\-\-config\fP \fIconfig_file\fP|\fB\-\-install\-sa\fP|\fB\-\-uninstall\-sa\fP|\fB\-\-check\-sa\fP|\fB\-\-load\-sa\fP]
.SH "DESCRIPTION"
.sp
\fByabai\fP is a tiling window manager for macOS based on binary space partitioning.
.SH "OPTIONS"
.sp
\fB\-v\fP, \fB\-\-version\fP
.RS 4
Print the version and exit.
.RE
.sp
\fB\-V\fP, \fB\-\-verbose\fP
.RS 4
Output debug information to stdout.
.RE
.sp
\fB\-m\fP, \fB\-\-message\fP \fI<msg>\fP
.RS 4
Send message to a running instance of yabai.
.RE
.sp
\fB\-c\fP, \fB\-\-config\fP \fI<config_file>\fP
.RS 4
Use the specified configuration file.
.RE
.sp
\fB\-\-install\-sa\fP
.RS 4
Install scripting\-addition. Must be run as root.
.br
Path is /System/Library/ScriptingAdditions on macOS High Sierra, and /Library/ScriptingAdditions on Mojave and newer.
.RE
.sp
\fB\-\-uninstall\-sa\fP
.RS 4
Uninstall scripting\-addition. Must be run as root.
.br
Path is /System/Library/ScriptingAdditions on macOS High Sierra, and /Library/ScriptingAdditions on Mojave and newer.
.RE
.sp
\fB\-\-check\-sa\fP
.RS 4
Returns a zero exit\-code if the latest version of the scripting\-addition is installed.
.RE
.sp
\fB\-\-load\-sa\fP
.RS 4
Loads the scripting\-addition into Dock.app.
.RE
.SH "DEFINITIONS"
.sp
.if n .RS 4
.nf
.fam C
REGEX       := \c
.URL "https://www.gnu.org/software/findutils/manual/html_node/find_html/posix_002dextended\-regular\-expression\-syntax.html" "POSIX extended regular expression syntax" ""

LABEL       := arbitrary string/text used as an identifier

LAYER       := below | normal | above

BOOL_SEL    := on | off

FLOAT_SEL   := 0 < <value> <= 1.0

RULE_SEL    := <index> | LABEL

SIGNAL_SEL  := <index> | LABEL

DIR_SEL     := north | east | south | west

STACK_SEL   := stack.prev | stack.next | stack.first | stack.last | stack.recent

WINDOW_SEL  := prev | next | first | last | recent | mouse | largest | smallest | STACK_SEL | DIR_SEL | <window id>

DISPLAY_SEL := prev | next | first | last | recent | mouse | DIR_SEL | <arrangement index (1\-based)>

SPACE_SEL   := prev | next | first | last | recent | mouse | <mission\-control index (1\-based)> | LABEL
.fam
.fi
.if n .RE
.SH "DOMAINS"
.SS "Config"
.SS "General Syntax"
.sp
yabai \-m config <global setting>
.RS 4
Get or set the value of <global setting>.
.RE
.sp
yabai \-m config [\-\-space \fI<SPACE_SEL>\fP] <space setting>
.RS 4
Get or set the value of <space setting>.
.RE
.SS "Global Settings"
.sp
\fBdebug_output\fP [\fI<BOOL_SEL>\fP]
.RS 4
Enable output of debug information to stdout.
.RE
.sp
\fBexternal_bar\fP [\fI<main|all|off>:<top_padding>:<bottom_padding>\fP]
.RS 4
Specify top and bottom padding for a potential custom bar that you may be running.
.br
\fImain\fP: Apply the given padding only to spaces located on the main display.
.br
\fIall\fP:  Apply the given padding to all spaces regardless of their display.
.br
\fIoff\fP:  Do not apply any special padding.
.RE
.sp
\fBmouse_follows_focus\fP [\fI<BOOL_SEL>\fP]
.RS 4
When focusing a window, put the mouse at its center.
.RE
.sp
\fBfocus_follows_mouse\fP [\fIautofocus|autoraise|off\fP]
.RS 4
Automatically focus the window under the mouse.
.RE
.sp
\fBwindow_origin_display\fP [\fIdefault|focused|cursor\fP]
.RS 4
Specify which display a newly created window should be managed in.
.br
\fIdefault\fP: The display in which the window is created (standard macOS behaviour).
.br
\fIfocused\fP: The display that has focus when the window is created.
.br
\fIcursor\fP: The display that currently holds the mouse cursor.
.RE
.sp
\fBwindow_placement\fP [\fIfirst_child|second_child\fP]
.RS 4
Specify whether managed windows should become the first or second leaf\-node.
.RE
.sp
\fBwindow_topmost\fP [\fI<BOOL_SEL>\fP]
.RS 4
Make floating windows stay on top.
.RE
.sp
\fBwindow_shadow\fP [\fI<BOOL_SEL>|float\fP]
.RS 4
Draw shadow for windows.
.RE
.sp
\fBwindow_opacity\fP [\fI<BOOL_SEL>\fP]
.RS 4
Enable opacity for windows.
.RE
.sp
\fBwindow_opacity_duration\fP [\fI<floating point number>\fP]
.RS 4
Duration of transition between active / normal opacity.
.RE
.sp
\fBwindow_animation_duration\fP [\fI<floating point number>\fP]
.RS 4
Duration of window frame animation. If 0.0, the change in dimension is not animated.
.RE
.sp
\fBactive_window_opacity\fP [\fI<FLOAT_SEL>\fP]
.RS 4
Opacity of the focused window.
.RE
.sp
\fBnormal_window_opacity\fP [\fI<FLOAT_SEL>\fP]
.RS 4
Opacity of an unfocused window.
.RE
.sp
\fBwindow_border\fP [\fI<BOOL_SEL>\fP]
.RS 4
Draw border for windows.
.RE
.sp
\fBwindow_border_width\fP [\fI<even integer number>\fP]
.RS 4
Width of window borders. If the given width is an odd number, it will be incremented by 1.
.RE
.sp
\fBactive_window_border_color\fP [\fI0xAARRGGBB\fP]
.RS 4
Color of the border of the focused window.
.RE
.sp
\fBnormal_window_border_color\fP [\fI0xAARRGGBB\fP]
.RS 4
Color of the border of an unfocused window.
.RE
.sp
\fBinsert_feedback_color\fP [\fI0xAARRGGBB\fP]
.RS 4
Color of the \fBwindow \-\-insert\fP message selection.
.RE
.sp
\fBsplit_ratio\fP [\fI<FLOAT_SEL>\fP]
.RS 4
Default split ratio.
.RE
.sp
\fBauto_balance\fP [\fI<BOOL_SEL>\fP]
.RS 4
Balance the window tree upon change, so that all windows occupy the same area.
.RE
.sp
\fBmouse_modifier\fP [\fIcmd|alt|shift|ctrl|fn\fP]
.RS 4
Keyboard modifier used for moving and resizing windows.
.RE
.sp
\fBmouse_action1\fP [\fImove|resize\fP]
.RS 4
Action performed when pressing \fImouse_modifier\fP + \fIbutton1\fP.
.RE
.sp
\fBmouse_action2\fP [\fImove|resize\fP]
.RS 4
Action performed when pressing \fImouse_modifier\fP + \fIbutton2\fP.
.RE
.sp
\fBmouse_drop_action\fP [\fIswap|stack\fP]
.RS 4
Action performed when a bsp\-managed window is dropped in the center of some other bsp\-managed window.
.RE
.SS "Space Settings"
.sp
\fBlayout\fP [\fIbsp|stack|float\fP]
.RS 4
Set the layout of the selected space.
.RE
.sp
\fBtop_padding\fP [\fI<integer number>\fP]
.RS 4
Padding added at the upper side of the selected space.
.RE
.sp
\fBbottom_padding\fP [\fI<integer number>\fP]
.RS 4
Padding added at the lower side of the selected space.
.RE
.sp
\fBleft_padding\fP [\fI<integer number>\fP]
.RS 4
Padding added at the left side of the selected space.
.RE
.sp
\fBright_padding\fP [\fI<integer number>\fP]
.RS 4
Padding added at the right side of the selected space.
.RE
.sp
\fBwindow_gap\fP [\fI<integer number>\fP]
.RS 4
Size of the gap that separates windows for the selected space.
.RE
.SS "Display"
.SS "General Syntax"
.sp
yabai \-m display [\fI<DISPLAY_SEL\fP>] \fI<COMMAND>\fP
.SS "COMMAND"
.sp
\fB\-\-focus\fP \fI<DISPLAY_SEL>\fP
.RS 4
Focus the given display.
.RE
.SS "Space"
.SS "General Syntax"
.sp
yabai \-m space [\fI<SPACE_SEL>\fP] \fI<COMMAND>\fP
.SS "COMMAND"
.sp
\fB\-\-focus\fP \fI<SPACE_SEL>\fP
.RS 4
Focus the given space.
.RE
.sp
\fB\-\-create\fP  [\fI<SPACE_SEL>\fP]
.RS 4
Create a new space on the display of the given space. If none specified, use the selected space instead.
.RE
.sp
\fB\-\-destroy\fP [\fI<SPACE_SEL>\fP]
.RS 4
Remove the given space. If none specified, use the selected space instead.
.RE
.sp
\fB\-\-move\fP \fI<SPACE_SEL>\fP
.RS 4
Move position of the selected space to the position of the given space.
.br
The selected space and given space must both belong to the same display.
.RE
.sp
\fB\-\-swap\fP \fI<SPACE_SEL>\fP
.RS 4
Swap the selected space with the given space.
.br
The selected space and given space must both belong to the same display.
.RE
.sp
\fB\-\-display\fP \fI<DISPLAY_SEL>\fP
.RS 4
Send the selected space to the given display.
.RE
.sp
\fB\-\-balance\fP [\fIx\-axis|y\-axis\fP]
.RS 4
Adjust the split ratios on the selected space so that all windows along the given axis occupy the same area. If no axis is specified, use both.
.RE
.sp
\fB\-\-mirror\fP \fIx\-axis|y\-axis\fP
.RS 4
Flip the tree of the selected space along the given axis.
.RE
.sp
\fB\-\-rotate\fP \fI90|180|270\fP
.RS 4
Rotate the tree of the selected space.
.RE
.sp
\fB\-\-padding\fP \fIabs|rel:<top>:<bottom>:<left>:<right>\fP
.RS 4
Padding added at the sides of the selected space.
.RE
.sp
\fB\-\-gap\fP \fIabs|rel:<gap>\fP
.RS 4
Size of the gap that separates windows on the selected space.
.RE
.sp
\fB\-\-toggle\fP \fIpadding|gap|mission\-control|show\-desktop\fP
.RS 4
Toggle space setting on or off for the selected space.
.RE
.sp
\fB\-\-layout\fP \fIbsp|stack|float\fP
.RS 4
Set the layout of the selected space.
.RE
.sp
\fB\-\-label\fP [\fI<LABEL>\fP]
.RS 4
Label the selected space, allowing that label to be used as an alias in commands that take a \f(CRSPACE_SEL\fP parameter.
.br
If the command is called without an argument it will try to remove a previously assigned label.
.RE
.SS "Window"
.SS "General Syntax"
.sp
yabai \-m window [\fI<WINDOW_SEL>\fP] \fI<COMMAND>\fP
.SS "COMMAND"
.sp
\fB\-\-focus\fP [\fI<WINDOW_SEL>\fP]
.RS 4
Focus the given window. If none specified, focus the selected window instead.
.RE
.sp
\fB\-\-swap\fP \fI<WINDOW_SEL>\fP
.RS 4
Swap position of the selected window and the given window.
.RE
.sp
\fB\-\-warp\fP \fI<WINDOW_SEL>\fP
.RS 4
Re\-insert the selected window, splitting the given window.
.RE
.sp
\fB\-\-stack\fP \fI<WINDOW_SEL>\fP
.RS 4
Stack the given window on top of the selected window.
.br
Any kind of warp operation performed on a stacked window will unstack it.
.RE
.sp
\fB\-\-insert\fP \fI<DIR_SEL>|stack\fP
.RS 4
Set the splitting mode of the selected window.
.br
If the current splitting mode matches the selected mode, the action will be undone.
.RE
.sp
\fB\-\-grid\fP \fI<rows>:<cols>:<start\-x>:<start\-y>:<width>:<height>\fP
.RS 4
Set the frame of the selected window based on a self\-defined grid.
.RE
.sp
\fB\-\-move\fP \fIabs|rel:<dx>:<dy>\fP
.RS 4
If type is \fIrel\fP the selected window is moved by \fIdx\fP pixels horizontally and \fIdy\fP pixels vertically, otherwise \fIdx\fP and \fIdy\fP will become its new position.
.RE
.sp
\fB\-\-resize\fP \fItop|left|bottom|right|top_left|top_right|bottom_right|bottom_left|abs:<dx>:<dy>\fP
.RS 4
Resize the selected window by moving the given handle \fIdx\fP pixels horizontally and \fIdy\fP pixels vertically. If handle is \fIabs\fP the new size will be \fIdx\fP width and \fIdy\fP height and cannot be used on managed windows.
.RE
.sp
\fB\-\-ratio\fP \fIrel|abs:<dr>\fP
.RS 4
If type is \fIrel\fP the split ratio of the selected window is changed by \fIdr\fP, otherwise \fIdr\fP will become the new split ratio. A positive/negative delta will increase/decrease the size of the left\-child.
.RE
.sp
\fB\-\-toggle\fP \fIfloat|sticky|topmost|pip|shadow|border|split|zoom\-parent|zoom\-fullscreen|native\-fullscreen|expose\fP
.RS 4
Toggle the given property of the selected window.
.RE
.sp
\fB\-\-layer\fP \fI<LAYER>\fP
.RS 4
Set the stacking layer of the selected window.
.RE
.sp
\fB\-\-opacity\fP \fI<floating point number>\fP
.RS 4
Set the opacity of the selected window. The window will no longer be eligible for automatic change in opacity upon focus change.
.br
Specify the value \fI0.0\fP to reset back to full opacity OR have it be automatically managed through focus changes.
.RE
.sp
\fB\-\-display\fP \fI<DISPLAY_SEL>\fP
.RS 4
Send the selected window to the given display.
.RE
.sp
\fB\-\-space\fP \fI<SPACE_SEL>\fP
.RS 4
Send the selected window to the given space.
.RE
.sp
\fB\-\-minimize\fP [\fI<WINDOW_SEL>\fP]
.RS 4
Minimizes the given window. If none specified, minimize the selected window instead. Only works on windows that provide a minimize button in its titlebar.
.RE
.sp
\fB\-\-deminimize\fP \fI<WINDOW_SEL>\fP
.RS 4
Restores the given window, if it is minimized. The window will only get focus if the owning application has focus.
.br
Note that you can also \fI\-\-focus\fP a minimized window to restore it as the focused window.
.RE
.sp
\fB\-\-close\fP [\fI<WINDOW_SEL>\fP]
.RS 4
Closes the given window. If none specified, close the selected window instead. Only works on windows that provide a close button in its titlebar.
.RE
.SS "Query"
.SS "General Syntax"
.sp
yabai \-m query \fI<COMMAND>\fP [\fI<ARGUMENT>\fP]
.SS "COMMAND"
.sp
\fB\-\-displays\fP
.RS 4
Retrieve information about displays.
.RE
.sp
\fB\-\-spaces\fP
.RS 4
Retrieve information about spaces.
.RE
.sp
\fB\-\-windows\fP
.RS 4
Retrieve information about windows.
.RE
.SS "ARGUMENT"
.sp
\fB\-\-display\fP [\fI<DISPLAY_SEL>\fP]
.RS 4
Constrain matches to the selected display.
.RE
.sp
\fB\-\-space\fP [\fI<SPACE_SEL>\fP]
.RS 4
Constrain matches to the selected space.
.RE
.sp
\fB\-\-window\fP [\fI<WINDOW_SEL>\fP]
.RS 4
Constrain matches to the selected window.
.RE
.SS "DATAFORMAT"
.sp
DISPLAY
.sp
.if n .RS 4
.nf
.fam C
{
    "id": number,
    "uuid": string,
    "index": number,
    "frame": object {
        "x": number,
        "y": number,
        "w": number,
        "h": number
    },
    "spaces": array of number
}
.fam
.fi
.if n .RE
.sp
SPACE
.sp
.if n .RS 4
.nf
.fam C
{
    "id": number,

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
- Implemented support for animating window move/resize operations (*config window_animation_duration*)[#148](https://github.com/koekeishiya/yabai/issues/148)

### Changed
- Implemented a workaround to support *window_opacity_duration*, bypassing the Apple bug [#1406](https://github.com/koekeishiya/yabai/issues/1406)

## [4.0.4] - 2022-09-08
### Changed
- Fix null-deref when an application is spawned with a non-standard window [#1399](https://github.com/koekeishiya/yabai/issues/1399)

## [4.0.3] - 2022-09-07
### Changed
- Allow combining commands for *config*, *space*, and *window* domains [#1371](https://github.com/koekeishiya/yabai/issues/1371)
- Fallback to using the AX API for *mouse_action move* if the scripting-addition is not available [#1376](https://github.com/koekeishiya/yabai/issues/1376)
- Add *role* and *subrole* filter to window rules [#1398](https://github.com/koekeishiya/yabai/issues/1398)

## [4.0.2] - 2022-08-24
### Changed
- Fixed an issue that in rare occasions caused yabai to freeze when focusing an inactve space on an inactive display [#1309](https://github.com/koekeishiya/yabai/issues/1309).
- Fixed an issue that caused a window to incorrectly become focused when assigned to a space through rules [#1370](https://github.com/koekeishiya/yabai/issues/1370)
- Fixed an issue that caused windows to snap back to its original position when moved between displays (to an inactive space) using mission-control [#820](https://github.com/koekeishiya/yabai/issues/820)

## [4.0.1] - 2022-05-17
### Changed
- The scripting-addition will now also remove the space switch animation when using cmd+tab, clicking on an item in the Dock, and using the numeric macOS mission-control keyboard shortcuts [#1235](https://github.com/koekeishiya/yabai/issues/1235)
- Improved logic used to determine the target window in a given direction [#1220](https://github.com/koekeishiya/yabai/issues/1220)
- Improve behaviour of *focus_follows_mouse autoraise*, preventing a window from being raised if it would occlude some other **floating window** [#1246](https://github.com/koekeishiya/yabai/issues/1246)
- The rule option *mouse_follows_focus* should now work properly (values were inverted) [#1267](https://github.com/koekeishiya/yabai/issues/1267)
- Remove minor shadow artifact from border windows [#1056](https://github.com/koekeishiya/yabai/issues/1056)

## [4.0.0] - 2022-03-16
### Added
- New config *window_origin_display* to specify which display a window should become managed at upon creation [#951](https://github.com/koekeishiya/yabai/issues/951)

### Changed
- Window borders no longer require SIP to be disabled [#1054](https://github.com/koekeishiya/yabai/issues/1054)
- WINDOW_SEL *prev* should correctly identify the correct window in a nested tree [#1114](https://github.com/koekeishiya/yabai/issues/1114)
- Fixed an issue with the way unix sockets were handled that would cause an incoming connection to drop in rare occasions [#1107](https://github.com/koekeishiya/yabai/issues/1107)
- Update scripting addition to support macOS 12.0.0 -> 12.3 [#1054](https://github.com/koekeishiya/yabai/issues/1054)
- Fixed an issue that would cause the target window to snap back to its previous position when moved between displays using the cursor inside mission-control [#820](https://github.com/koekeishiya/yabai/issues/820)
- Properly drain autoreleased objects (from Apple frameworks) [#751](https://github.com/koekeishiya/yabai/issues/751)
- Detect and manage windows that are moved into the first space of a display when a space with active windows on it is destroyed [#813](https://github.com/koekeishiya/yabai/issues/813)
- The command *space --balance* now takes an optional axis as its argument [#985](https://github.com/koekeishiya/yabai/issues/985)
- Malformed windows that are managed through rules should be eligible for window alpha, focus follows mouse, and window borders [#788](https://github.com/koekeishiya/yabai/issues/788)
- Improve behaviour of *focus_follows_mouse autoraise*, preventing a window from being raised if it would occlude some other window. autoraise is temporarily disabled while a menu is opened. [#407](https://github.com/koekeishiya/yabai/issues/407)
- Work around broken Apple code for retrieving menubar dimensions on Apple Silicon M1 [#793](https://github.com/koekeishiya/yabai/issues/793)
- Rework query attributes and define dataformat as part of the API [#775](https://github.com/koekeishiya/yabai/issues/775)
- Properly clear space.last-window and space.first-window query attributes when the last window is made floating [#786](https://github.com/koekeishiya/yabai/issues/786)
- Reworked signal system; events are no longer coupled 1-1 with observed system events.
Some events are now eligible for a new filter, *active*, only triggering for the application/window with key-focus.
The *window_focused* signal is now triggered when the key-window changes, regardless of whether its application is frontmost or not.
The *window_created* signal is now triggered for windows that are implicitly created at application launch.
The *window_destroyed* signal is now triggered for windows that are implicitly destroyed at application exit; it is also eligible for *app* filter [#581](https://github.com/koekeishiya/yabai/issues/581)

### Removed
- The following signals have been removed: *mouse_up*, *mouse_down*, *mouse_dragged*, *mouse_moved*, *mission_control_check_for_exit*, *menu_opened*, *system_woke*, *daemon_message*

## [3.3.10] - 2021-05-26
### Changed
- Update scripting addition to support macOS 10.15.7 Supplemental Update

## [3.3.9] - 2021-05-25
### Changed
- Update scripting addition to support macOS 11.4

## [3.3.8] - 2021-04-28
### Changed
- Update scripting addition to support macOS 11.3

## [3.3.7] - 2021-02-02

src/yabai.c
#define SA_SOCKET_PATH_FMT      "/tmp/yabai-sa_%s.socket"

#define SOCKET_PATH_FMT         "/tmp/yabai_%s.socket"
#define LCFILE_PATH_FMT         "/tmp/yabai_%s.lock"

#define CLIENT_OPT_LONG         "--message"
#define CLIENT_OPT_SHRT         "-m"

#define DEBUG_VERBOSE_OPT_LONG  "--verbose"
#define DEBUG_VERBOSE_OPT_SHRT  "-V"
#define VERSION_OPT_LONG        "--version"
#define VERSION_OPT_SHRT        "-v"
#define CONFIG_OPT_LONG         "--config"
#define CONFIG_OPT_SHRT         "-c"

#define SCRPT_ADD_INSTALL_OPT   "--install-sa"
#define SCRPT_ADD_UNINSTALL_OPT "--uninstall-sa"
#define SCRPT_ADD_CHECK_OPT     "--check-sa"
#define SCRPT_ADD_LOAD_OPT      "--load-sa"

#define MAJOR  4
#define MINOR  0
#define PATCH  4

struct event_loop g_event_loop;
void *g_workspace_context;
struct process_manager g_process_manager;
struct display_manager g_display_manager;
struct space_manager g_space_manager;
struct window_manager g_window_manager;
struct mouse_state g_mouse_state;
struct event_tap g_event_tap;
int g_normal_window_level;
int g_floating_window_level;
int g_connection;

struct signal *g_signal_event[SIGNAL_TYPE_COUNT];
struct memory_pool g_signal_storage;
int g_mission_control_active;
char g_sa_socket_file[MAXLEN];
char g_socket_file[MAXLEN];
char g_config_file[4096];
char g_lock_file[MAXLEN];
bool g_verbose;

struct mach_server g_mach_server;

static int client_send_message(int argc, char **argv)
{
    if (argc <= 1) {
        error("yabai-msg: no arguments given! abort..\n");
    }

    char *user = getenv("USER");
    if (!user) {
        error("yabai-msg: 'env USER' not set! abort..\n");
    }

    int message_length = argc;
    int argl[argc];

    for (int i = 1; i < argc; ++i) {
        argl[i] = strlen(argv[i]);
        message_length += argl[i] + 1;
    }

    char *message = malloc(sizeof(int)+message_length + 1);
    char *temp = sizeof(int)+message;

    memcpy(message, &message_length, sizeof(int));
    for (int i = 1; i < argc; ++i) {
        memcpy(temp, argv[i], argl[i]);
        temp += argl[i];
        *temp++ = '\0';
    }
    *temp++ = '\0';

    char* rsp = mach_send_message(mach_get_bs_port(),
                                  message,
                                  message_length,
                                  true               );

    int result = EXIT_SUCCESS;
    FILE *output = stdout;

    if (!rsp) return result;
    if (rsp[0] == FAILURE_MESSAGE[0]) {
        result = EXIT_FAILURE;
        output = stderr;
        fprintf(output, "%s", rsp + 1);
    } else {
        fprintf(output, "%s", rsp);
    }
    return result;
}

static void acquire_lockfile(void)
{
    int handle = open(g_lock_file, O_CREAT | O_WRONLY, 0600);
    if (handle == -1) {
        error("yabai: could not create lock-file! abort..\n");
    }

    struct flock lockfd = {
        .l_start  = 0,
        .l_len    = 0,
        .l_pid    = getpid(),
        .l_type   = F_WRLCK,
        .l_whence = SEEK_SET
    };

    if (fcntl(handle, F_SETLK, &lockfd) == -1) {
        error("yabai: could not acquire lock-file! abort..\n");
    }
}

static bool get_config_file(char *restrict filename, char *restrict buffer, int buffer_size)
{
    char *xdg_home = getenv("XDG_CONFIG_HOME");
    if (xdg_home && *xdg_home) {
        snprintf(buffer, buffer_size, "%s/yabai/%s", xdg_home, filename);
        if (file_exists(buffer)) return true;
    }

    char *home = getenv("HOME");
    if (!home) return false;

    snprintf(buffer, buffer_size, "%s/.config/yabai/%s", home, filename);
    if (file_exists(buffer)) return true;

    snprintf(buffer, buffer_size, "%s/.%s", home, filename);
    return file_exists(buffer);
}

static void exec_config_file(void)
{
    if (!*g_config_file && !get_config_file("yabairc", g_config_file, sizeof(g_config_file))) {
        notify("configuration", "could not locate config file..");
        return;
    }

    if (!file_exists(g_config_file)) {
        notify("configuration", "file '%s' does not exist..", g_config_file);
        return;
    }

    if (!ensure_executable_permission(g_config_file)) {
        notify("configuration", "could not set the executable permission bit for '%s'", g_config_file);
        return;
    }

    int pid = fork();
    if (pid == 0) {
        char *exec[] = { "/usr/bin/env", "sh", "-c", g_config_file, NULL};
        exit(execvp(exec[0], exec));
    } else if (pid == -1) {
        notify("configuration", "failed to execute file '%s'", g_config_file);
    }
}

#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"
static inline void init_misc_settings(void)
{
    char *user = getenv("USER");
    if (!user) {
        error("yabai: 'env USER' not set! abort..\n");
    }

    snprintf(g_sa_socket_file, sizeof(g_sa_socket_file), SA_SOCKET_PATH_FMT, user);
    snprintf(g_socket_file, sizeof(g_socket_file), SOCKET_PATH_FMT, user);
    snprintf(g_lock_file, sizeof(g_lock_file), LCFILE_PATH_FMT, user);

    NSApplicationLoad();
    signal(SIGCHLD, SIG_IGN);
    signal(SIGPIPE, SIG_IGN);
    CGSetLocalEventsSuppressionInterval(0.0f);
    CGEnableEventStateCombining(false);
    g_connection = SLSMainConnectionID();
    g_normal_window_level   = CGWindowLevelForKey(LAYER_NORMAL);
    g_floating_window_level = CGWindowLevelForKey(LAYER_ABOVE);

#if 0
    hook_nsobject_autorelease();
    hook_autoreleasepool_drain();
    hook_autoreleasepool_release();
#endif
}
#pragma clang diagnostic pop

static CONNECTION_CALLBACK(connection_handler)
{
    if (type == 1204) {
        event_loop_post(&g_event_loop, MISSION_CONTROL_ENTER, NULL, 0, NULL);
    } else if (type == 806) {
        event_loop_post(&g_event_loop, SLS_WINDOW_MOVED, (void *) (intptr_t) (*(uint32_t *) data), 0, NULL);
    } else if (type == 807) {
        event_loop_post(&g_event_loop, SLS_WINDOW_RESIZED, (void *) (intptr_t) (* (uint32_t *) data), 0, NULL);
    } else if (type == 808) {
        event_loop_post(&g_event_loop, SLS_WINDOW_ORDER_CHANGED, (void *) (intptr_t) (* (uint32_t *) data), 0, NULL);
    } else if (type == 815) {
        event_loop_post(&g_event_loop, SLS_WINDOW_IS_VISIBLE, (void *) (intptr_t) (* (uint32_t *) data), 0, NULL);
    } else if (type == 816) {
        event_loop_post(&g_event_loop, SLS_WINDOW_IS_INVISIBLE, (void *) (intptr_t) (* (uint32_t *) data), 0, NULL);
    }
}

static void parse_arguments(int argc, char **argv)
{
    if ((string_equals(argv[1], VERSION_OPT_LONG)) ||
        (string_equals(argv[1], VERSION_OPT_SHRT))) {
        fprintf(stdout, "yabai-v%d.%d.%d\n", MAJOR, MINOR, PATCH);
        exit(EXIT_SUCCESS);
    }

    if ((string_equals(argv[1], CLIENT_OPT_LONG)) ||
        (string_equals(argv[1], CLIENT_OPT_SHRT))) {
        exit(client_send_message(argc-1, argv+1));
    }

    if (string_equals(argv[1], SCRPT_ADD_INSTALL_OPT)) {
        exit(scripting_addition_install());
    }

    if (string_equals(argv[1], SCRPT_ADD_UNINSTALL_OPT)) {
        exit(scripting_addition_uninstall());
    }

    if (string_equals(argv[1], SCRPT_ADD_CHECK_OPT)) {
        exit(scripting_addition_check());
    }

    if (string_equals(argv[1], SCRPT_ADD_LOAD_OPT)) {
        exit(scripting_addition_load());
    }

    for (int i = 1; i < argc; ++i) {
        char *opt = argv[i];

        if ((string_equals(opt, DEBUG_VERBOSE_OPT_LONG)) ||
            (string_equals(opt, DEBUG_VERBOSE_OPT_SHRT))) {
            g_verbose = true;
        } else if ((string_equals(opt, CONFIG_OPT_LONG)) ||
                   (string_equals(opt, CONFIG_OPT_SHRT))) {
            char *val = i < argc - 1 ? argv[++i] : NULL;
            if (!val) error("yabai: option '%s|%s' requires an argument!\n", CONFIG_OPT_LONG, CONFIG_OPT_SHRT);
            snprintf(g_config_file, sizeof(g_config_file), "%s", val);
        } else {
            error("yabai: '%s' is not a valid option!\n", opt);
        }
    }
}

int main(int argc, char **argv)
{
    if (argc > 1) {
        parse_arguments(argc, argv);
    }

    if (is_root()) {
        error("yabai: running as root is not allowed! abort..\n");
    }

    if (!ax_privilege()) {
        error("yabai: could not access accessibility features! abort..\n");
    }

    init_misc_settings();
    acquire_lockfile();

    if (!space_manager_has_separate_spaces()) {
        error("yabai: 'display has separate spaces' is disabled! abort..\n");
    }

    if (!event_loop_init(&g_event_loop)) {
        error("yabai: could not initialize event_loop! abort..\n");
    }

    if (!memory_pool_init(&g_signal_storage, KILOBYTES(128))) {
        error("yabai: could not allocate memory for event_signal! abort..\n");
    }

    if (!ts_init(MEGABYTES(4))) {
        error("yabai: could not allocate temporary storage! abort..\n");
    }

    process_manager_init(&g_process_manager);
    workspace_event_handler_init(&g_workspace_context);
    space_manager_init(&g_space_manager);
    window_manager_init(&g_window_manager);
    mouse_state_init(&g_mouse_state);

    event_loop_begin(&g_event_loop);
    display_manager_begin(&g_display_manager);
    space_manager_begin(&g_space_manager);
    window_manager_begin(&g_space_manager, &g_window_manager);
    process_manager_begin(&g_process_manager);
    workspace_event_handler_begin(&g_workspace_context);
    event_tap_begin(&g_event_tap, EVENT_MASK_MOUSE, mouse_handler);

    if (workspace_is_macos_monterey()) {
        mission_control_observe();
    } else {
        SLSRegisterConnectionNotifyProc(g_connection, connection_handler, 1204, NULL);
    }

    SLSRegisterConnectionNotifyProc(g_connection, connection_handler, 806, NULL);
    SLSRegisterConnectionNotifyProc(g_connection, connection_handler, 807, NULL);
    SLSRegisterConnectionNotifyProc(g_connection, connection_handler, 808, NULL);
    SLSRegisterConnectionNotifyProc(g_connection, connection_handler, 815, NULL);
    SLSRegisterConnectionNotifyProc(g_connection, connection_handler, 816, NULL);

    if (!mach_server_begin(&g_mach_server, mach_message_handler))
      error("yabai: could not initialize daemon! abort..\n");

    if (!workspace_is_macos_monterey() && !workspace_is_macos_bigsur()) {
        if (scripting_addition_is_installed()) {
            scripting_addition_load();
        } else {
            notify("scripting-addition", "payload is not installed, some features will not work!");
        }
    }

    exec_config_file();

    for (;;) {
        NSAutoreleasePool *pool = [[NSAutoreleasePool alloc] init];
        CFRunLoopRunResult result = CFRunLoopRunInMode(kCFRunLoopDefaultMode, 300, true);
        [pool drain];

        if (result == kCFRunLoopRunFinished || result == kCFRunLoopRunStopped) break;
    }

    return 0;
}
